harness/determined/common/experimental/determined.py

```diff
@@ -207,27 +177,75 @@ def create_experiment(
-
-    def get_experiment(self, experiment_id: int) -> experiment.ExperimentReference:
+    def get_experiment(self, experiment_id: int) -> experiment.Experiment:
         """
-        Get the :class:`~determined.experimental.ExperimentReference` representing the
+        Get the :class:`~determined.experimental.Experiment` representing the
         experiment with the provided experiment ID.
         """
-        return experiment.ExperimentReference(
-            experiment_id,
-            self._session,
+        resp = bindings.get_GetExperiment(session=self._session, experimentId=experiment_id)
+        return experiment.Experiment._from_bindings(resp.experiment, self._session)
+
+    def list_experiments(
+        self,
+        sort_by: Optional[experiment.ExperimentSortBy] = None,
+        order_by: Optional[experiment.ExperimentOrderBy] = None,
+        experiment_ids: Optional[List[int]] = None,
+        labels: Optional[List[str]] = None,
+        users: Optional[List[str]] = None,
+        states: Optional[List[experiment.ExperimentState]] = None,
+        name: Optional[str] = None,
+        project_id: Optional[int] = None,
+    ) -> List[experiment.Experiment]:
+        """Get a list of experiments (:class:`~determined.experimental.Experiment`).
+
+        Arguments:
+            sort_by: Which field to sort by. See
+                :class:`~determined.experimental.ExperimentSortBy`.
+            order_by: Whether to sort in ascending or descending order. See
+                :class:`~determined.experimental.ExperimentOrderBy`.
+            name: If this parameter is set, experiments will be filtered to only include those
+                with names matching this parameter.
+            experiment_ids: Only return experiments with these IDs.
+            labels: Only return experiments with a label in this list.
+            users: Only return experiments belonging to these users. Defaults to all users.
+            states: Only return experiments that are in these states.
+            project_id: Only return experiments associated with this project ID.
+
+        Returns:
+            A list of experiments.
+        """
+
+        def get_with_offset(offset: int) -> bindings.v1GetExperimentsResponse:
+            return bindings.get_GetExperiments(
+                session=self._session,
+                sortBy=sort_by and sort_by._to_bindings() or None,
+                orderBy=order_by and order_by._to_bindings() or None,
+                archived=None,
+                description=None,
+                labels=labels,
+                experimentIdFilter_incl=experiment_ids,
+                offset=offset,
+                limit=None,
+                name=name,
+                states=[state._to_bindings() for state in states] if states else None,
+                users=users,
+                projectId=project_id,
+            )
+
+        bindings_exps: Iterable[bindings.v1Experiment] = itertools.chain.from_iterable(
```

Comment:
what does itertools do here that a plain iterator doesn't do?

```python
binding_exps = (r.experiments for r in api.read_paginated(get_with_offset))
```
?

Reply:
(same question applies in other places I think)

Reply:
I would also expect that using a plain iterator means you don't need to specify the `Iterable[bindings.v1Experiment]`, since the primary design goal of `read_paginated` is to play nicely with mypy.  I _think_ the only reason you need an annotation is the involvement of itertools.chain.

Reply:
Your question is a question I have almost every dang time I try to use `api.read_paginated`. The type of `binding_exps` in
```python
binding_exps = (r.experiments for r in api.read_paginated(get_with_offset))
```
is `Iterable[List[bindings.v1Experiment]]`, not `Iterable[bindings.v1Experiment]`. The flattening is what `itertools` does for us, and what the nested comprehension does in
```python
exps = (e for e in r.experiments for r in api.read_paginated(get_with_offset))
```
(or whatever, the nested syntax is _also_ hard for me to read, which is why I made this change).

That our bindings return pages instead of items is not human, `api.read_paginated` doesn't help very much. I've got a [draft PR](https://github.com/determined-ai/determined/pull/7955) to make it optionally flatten pages. It's _ok_, as evidenced by [this followup PR](https://github.com/determined-ai/determined/pull/7993) that actually uses the new option. Better, but the flattening really should be in the bindings, rather than in `api` (also, the introspective magic in the flattening `read_paginated` doesn't work in all cases -- there are several paginated endpoints that return both resource pages and extra, irrelevant information because who knows why).

You _should_ be able to get a raw response, complete with list and pagination for whatever advanced use case I can't even imagine. But every use case I've dealt with so far personally only wants the resources inside the lists, not the pages themselves.

It's also true that with this itertools solution that `mypy` can infer the type annotation, and that explicitly including it isn't necessary to get it to typecheck. I included it here because this is a new idiom to a common use case in the codebase, and I thought it might help people understand how it works.

Reply:
Ah, I see.

I recall that `api.read_paginated` was the best I could come up with to keep mypy happy and still have automatic paging.  I do think it's awkward to use. 

IMO, the bindings are _not_ there to be ergonomic (or "human") at all.  The SDK is there to be ergonomic.  The bindings are there to be strongly typed.  That's exactly why when we introduced the bindings we also said "bindings shall never be handed to end users, because generated bindings are always awkward to use".

That said, you could totally have the bindings generate paginated getters for each of the paginated responses.  That wouldn't be too hard.  It's our bindings generator, anyway.

For now, breaking mypy here feels like a mistake.  Type safety was literally the reason why we introduced generated bindings.

Feel free to consider this feedback non-blocking, my approval stands.

Reply:
ok, I didn't understand the last part of what you said, that it does work with mypy even without the annotation.

Nevermind!

Reply:
Chatted OOB, verified that mypy does infer the type of `binding_exps` correctly even without explicit annotation. So what's left is the tradeoff:
* explicit annotation is negative because it makes it look like we've written code mypy can't infer types from
* explicit annotation is positive because it gives the human reader a hint for understanding the code

Reply:
[Style guide](https://google.github.io/styleguide/pyguide.html) doesn't comment on the difference between needed and unneeded annotations (and, honestly, is a little behind the times wrt type checking at all). Closest I can find is:

> If an internal variable has a type that is hard or impossible to infer, specify its type with an annotated assignment

Since "hard" means hard for a human, I'm assuming that this is a comment about readability of the code.

So... I _do_ think the style guide is a little out of date, and I might be using it to rationalize a decision I want to make, but I think it at least leans pro-annotation in this case. I'd like to leave it in place.